fpga_interchange/site_router.cc

```diff
@@ -0,0 +1,753 @@
+                    expansion.remove_wire(consumed_wire);
+                }
+
+                // By removing that wire, this expansion now has no solutions!
+                if (expansion.completed_routes.empty()) {
+                    return false;
+                }
+            }
+        }
+
+        // Check if there are any more trival solutions.
+        completed_wires.clear();
+        newly_consumed_wires.clear();
+
+        for (auto &expansion_wire : wire_to_expansion) {
+            auto &expansion = *expansion_wire.second;
+            if (expansion.completed_routes.size() == 1) {
+                auto node = expansion.completed_routes.begin()->second;
+                if (!site_info->select_route(expansion.first_wire, node, expansion.net_for_wire,
+                                             &newly_consumed_wires)) {
+                    // Conflict!
+                    return false;
+                }
+
+                // Mark this expansion as done!
+                completed_wires.push_back(expansion_wire.first);
+            }
+        }
+
+        // Remove trival solutions from unsolved routing.
+        for (WireId wire : completed_wires) {
+            NPNR_ASSERT(wire_to_expansion.erase(wire) == 1);
+        }
+
+        // All expansions have been selected for!
+        if (wire_to_expansion.empty()) {
+            break;
+        }
+
+        // At least 1 trival solution was selected, re-prune.
+        if (!newly_consumed_wires.empty()) {
+            // Prune remaining solutions.
+            continue;
+        }
+
+        std::unordered_map<WireId, std::unordered_set<const NetInfo *>> wire_congestion;
+
+        for (auto &expansion_wire : wire_to_expansion) {
+            auto &expansion = *expansion_wire.second;
+
+            for (auto pair : expansion.completed_routes) {
+                auto node = pair.second;
+
+                do {
+                    wire_congestion[node->wire].emplace(expansion.net_for_wire);
+                    node = node->parent;
+                } while (node != RouteNode::Node());
+            }
+        }
+
+        for (auto &expansion_wire : wire_to_expansion) {
+            auto &expansion = *expansion_wire.second;
+
+            RouteNode::Node uncongestion_route;
+
+            for (auto pair : expansion.completed_routes) {
+                auto node = pair.second;
+                uncongestion_route = node;
+
+                do {
+                    if (wire_congestion[node->wire].size() > 1) {
+                        uncongestion_route = RouteNode::Node();
+                        break;
+                    }
+                    node = node->parent;
+                } while (node != RouteNode::Node());
+
+                if (uncongestion_route != RouteNode::Node()) {
+                    break;
+                }
+            }
+
+            if (uncongestion_route != RouteNode::Node()) {
+                // Select a trivally uncongestion route if possible.
```

Comment:
nitpick - these comments could do with a spelling check

Reply:
Indeed, fixed it up!